## Problem Statement

The first thing any visitor to India will take in — probably while staring out the window in awe as their aeroplane descends is the sheer size of this country. It is densely populated and patchworked with distinct neighbourhoods, each with its own culinary identity. It would take several lifetimes to get to know all of the street stands, holes in the wall, neighbourhood favourites, and high-end destinations in this city.
And for Indians dining out is and always will be a joyous occasion. Everyone has their own favourite restaurants in the city starting from the street food stall across the street to the 5-star restaurants in the heart of the city. Some are favourites because of the memory attached to it and some are favourites because of the fact that the place has a fantastic ambience. There are a lot of other factors as well which contribute to the likeness of the restaurants which in turn determines their popularity among masses. 

If you look at this from the business perspective for a restaurant, more popularity may mean more visits to the joint increasing the annual turnover of the restaurants. For any restaurant to survive and do well, the annual turnover of the restaurants has to be substantial. 

This problem takes a shot at predicting the annual turnover of a set of restaurants across India based on a set of variables given in the data set. This includes the data related to the restaurant such as location, opening date, cuisine type, themes etc. This also includes data pooled from different sources such as social media popularity index, Zomato ratings, etc. Lastly, it also adds a different flavour to the problem by looking at the Customer survey data as well as ratings provided by a mystery visitor data (audit done by a third party). 

## Data Dictionary

| S No. | Variable Name | Variable Description |
| ----- | ------------- | -------------------- |
| 1 | Registration Number | This is a restaurant ID |
| 2 | Annual Turnover | Annual revenue of the restaurant in INR |
| 3 | Cuisine | Type of cuisine served in the restaurant |
| 4 | City | City at which the restaurant is located |
| 5 | Restaurant Location | This variable determines whether the restaurant is located near a business hub or a party hub |
| 6 | Opening Day of Restaurant | Date of the opening of the restaurant |
| 7 | Facebook Popularity Quotient | "Out of 100, this is the popularity of the restaurant on Facebook" |
| 8 | Endoresed By | This variable tells us what kind of celebrity endorses the restaurant |
| 9 | Instagram Popularity Quotient | "Out of 100, this is the popularity of the restaurant on Instragram" |
| 10 | Fire Audit | "This variable tells us whether the fire audit is succesfull in the restaurant. Here 1 means appropriate fire safety is present, 0 means appropriate fire safety is not present" |
| 11 | Liquor License Obtained | "This variable tells us whether the restaurant has liquour license or not. 1 means liquor license is present, 0 means otherwise" |
| 12 | Situated in a Multi Complex | "This variable tells us whether the restaurant is situated in a multi complex. 1 means the restaurant is present in the multi complex, 0 means otherwise" |
| 13 | Dedicated Parking | "This variable tells us whether the restaurant has a dedicated parking space. 1 means dedicated parking space is present, 0 means otherwise" |
| 14 | Open Sitting Available | "This variable tells us whether the restaurant has opening sitting. 1 means opening sitting is present, 0 means otherwise" |
| 15 | Resturant Tier | This variable tells us what tier the restaurant belongs to. |
| 16 | Restaurant Type | This variable tells us the type of restaurant. |
| 17 | Restaurant Theme | This variable tells us the theme of the restaurant by which it is designed. |
| 18 | Restaurant Zomato Rating | "This variable tells us the Zomato rating of the restaurant on a scale of 1 to 5, 5 being the highest." |
| 19 | Restaurant City Tier | This variable tells us the tier that restaurant belongs to |
| 20 | Order Wait Time | "This variable rates the waiting time of the restaurant on a scale of 1 to 10, 10 being the highest" |
| 21 | Staff Responsivness | "This variable rates the staff responsive of the restaurant on a scale of 1 to 8, 8 being the highest" |
| 22 | Value for Money | "This variable rates the staff responsive of the restaurant on a scale of 1 to 7, 7 being the highest" |
| 23 | Hygiene Rating | "This is the Hygiene Rating of restuarent on a scale of 1 to 10,10 being the highest" |
| 24 | Food Rating | "This is the food Rating of restuarent on a scale of 1 to 10, 10 being the highest" |
| 25 | Overall Restaurant Rating | "This is the overall restaurant rating on a scale of 1 to 10, 10 being the highest" |
| 26 | Live Music Rating | "This variable gives an indication about the satisfacation from the Live musicon a scale of 1 to 10, 10 being the highest, NA means restuarant do not offer Live music" |
| 27 | Comedy Gigs Rating | "This variable gives an indication about the satisfacation from the Comedy Show on a scale of 1 to 6, 6 being the highest NA means restuarant do not offer any comedy gigs" | 
| 28 | Value Deals Rating | "This variable gives an indication about the satisfacation from the Value Deals on a scale of 1 to 7, 7 being the highest NA means restuarant do not offer any value deals" |
| 29 | Live Sports Rating | "This variable gives an indication about the satisfacation from the Live screening of Sports on a scale of 1 to 6, 6 being the highestNA means restuarant do not have live screening" |
| 30 | Ambience | "This variable gives us an indication about the ambience feel level rating of the restuarant on a scale of 0 to 10, 10 being the highest" |
| 31 | Lively | "This variable rates the lively atmosphere of the restaurant on a scale of 1 to 10, 10 being the highest" |
| 32 | Service | This variable gives us an indication about the service satisfaction level rating of the restuarant. Here Rating of 10 means highly Satisfied from the service and 0 means otherwise |
| 33 | Comfortablility | "This variable gives us an indication about the comfort level rating of the restuarent on a scale of 0 to 10, 10 being the highest" |
| 34 | Privacy | "This variable gives us an indication about the privacy level of the restuarant on a scale of 0 to 10, 10 being the highest" |

## Import Libraries

In [1]:
# to load and manipulate data
import pandas as pd
import numpy as np

# to visualize data
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# Import to get idea of skew prior to exploring data
from scipy.stats import skew
from scipy.stats.mstats import winsorize

# to split data into training and test sets
from sklearn.model_selection import train_test_split

# to tune different models
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestRegressor, StackingRegressor, VotingRegressor
from sklearn.metrics import mean_squared_error
from xgboost import DMatrix, cv
import xgboost as xgb
import statsmodels.api as sm
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import HuberRegressor
from sklearn.ensemble import AdaBoostRegressor
import lightgbm as lgb
from sklearn.linear_model import QuantileRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV, train_test_split
from sklearn.metrics import classification_report, confusion_matrix, precision_score

# to compute classification metrics
from sklearn.metrics import silhouette_score
import sklearn.metrics

# to scale the data using z-score
from sklearn.preprocessing import StandardScaler

# to perform k-means clustering and compute silhouette scores
from sklearn.cluster import KMeans
from sklearn.cluster import DBSCAN

# to perform t-SNE and PCA
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA
from scipy.stats import zscore
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.linear_model import ElasticNet
from sklearn.linear_model import ElasticNetCV

# Import fuzzywuzzy for fuzzy searching of cities
from fuzzywuzzy import process

# To ignore unnecessary warnings
import warnings
warnings.filterwarnings("ignore")
pd.set_option("display.max_rows", 200)

# to define a common seed value to be used throughout
RS=23

/home/jimmyb/repos/UTAustin_Turnover/.venv/lib/python3.12/site-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


## Data Overview

In [2]:
restaurants = pd.read_csv('data/Train_dataset_(1)_(1)_(2).csv')
data = restaurants.copy()
data.head()

Registration Number  Annual Turnover           Cuisine       City  \
0                60001         42000000      indian,irish  Bangalore   
1                60002         50000000      indian,irish     Indore   
2                60003         32500000   tibetan,italian    Chennai   
3                60004        110000000  turkish,nigerian    Gurgaon   
4                60005         20000000     irish,belgian    Manesar   

  Restaurant Location Opening Day of Restaurant  Facebook Popularity Quotient  \
0   Near Business Hub                14-02-2009                          84.3   
1      Near Party Hub                29-09-2008                          85.4   
2   Near Business Hub                30-07-2011                          85.0   
3      Near Party Hub                30-11-2008                          85.6   
4      Near Party Hub                22-02-2010                           NaN   

        Endorsed By  Instagram Popularity Quotient  Fire Audit  ...  \
0      Not Specific                           95.8           1  ...   
1  Tier A Celebrity                           85.0           1  ...   
2  Tier A Celebrity                           68.2           1  ...   
3  Tier A Celebrity                           83.6           0  ...   
4  Tier A Celebrity                           76.8           1  ...   

   Overall Restaurant Rating  Live Music Rating  Comedy Gigs Rating  \
0                       10.0                4.0                 NaN   
1                        9.0                NaN                 4.0   
2                        8.0                3.0                 NaN   
3                        9.0                6.0                 NaN   
4                        6.0                NaN                 2.0   

   Value Deals Rating  Live Sports Rating Ambience Lively  Service  \
0                 NaN                 NaN      8.0      8        6   
1                 NaN                 NaN      5.0      7        7   
2                 NaN                 NaN      7.0     10        5   
3                 NaN                 NaN      7.0      7        4   
4                 NaN                 NaN      NaN      6        2   

   Comfortablility  Privacy  
0                6        6  
1                3        8  
2                2        8  
3                3        5  
4                4        6  

[5 rows x 34 columns]

In [3]:
data.shape

(3493, 34)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3493 entries, 0 to 3492
Data columns (total 34 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Registration Number            3493 non-null   int64  
 1   Annual Turnover                3493 non-null   int64  
 2   Cuisine                        3493 non-null   object 
 3   City                           3493 non-null   object 
 4   Restaurant Location            3493 non-null   object 
 5   Opening Day of Restaurant      3493 non-null   object 
 6   Facebook Popularity Quotient   3394 non-null   float64
 7   Endorsed By                    3493 non-null   object 
 8   Instagram Popularity Quotient  3437 non-null   float64
 9   Fire Audit                     3493 non-null   int64  
 10  Liquor License Obtained        3493 non-null   int64  
 11  Situated in a Multi Complex    3493 non-null   int64  
 12  Dedicated Parking              3493 non-null   i

In [5]:
data.describe(include='all').T

count unique             top  freq  \
Registration Number            3493.0    NaN             NaN   NaN   
Annual Turnover                3493.0    NaN             NaN   NaN   
Cuisine                          3493     20   tibetan,greek   686   
City                             3493    297       Bangalore   553   
Restaurant Location              3493      2  Near Party Hub  2658   
Opening Day of Restaurant        3493   1778      27-12-2009    10   
Facebook Popularity Quotient   3394.0    NaN             NaN   NaN   
Endorsed By                      3493      3    Not Specific  1961   
Instagram Popularity Quotient  3437.0    NaN             NaN   NaN   
Fire Audit                     3493.0    NaN             NaN   NaN   
Liquor License Obtained        3493.0    NaN             NaN   NaN   
Situated in a Multi Complex    3493.0    NaN             NaN   NaN   
Dedicated Parking              3493.0    NaN             NaN   NaN   
Open Sitting Available         3493.0    NaN             NaN   NaN   
Resturant Tier                 3444.0    NaN             NaN   NaN   
Restaurant Type                  3493      4             Bar  3235   
Restaurant Theme                 3493     34         Arabian  1373   
Restaurant Zomato Rating       3493.0    NaN             NaN   NaN   
Restaurant City Tier           3493.0    NaN             NaN   NaN   
Order Wait Time                3493.0    NaN             NaN   NaN   
Staff Responsivness            3493.0    NaN             NaN   NaN   
Value for Money                3493.0    NaN             NaN   NaN   
Hygiene Rating                 3493.0    NaN             NaN   NaN   
Food Rating                    3493.0    NaN             NaN   NaN   
Overall Restaurant Rating      3281.0    NaN             NaN   NaN   
Live Music Rating              2728.0    NaN             NaN   NaN   
Comedy Gigs Rating             1010.0    NaN             NaN   NaN   
Value Deals Rating              786.0    NaN             NaN   NaN   
Live Sports Rating              205.0    NaN             NaN   NaN   
Ambience                       3468.0    NaN             NaN   NaN   
Lively                         3493.0    NaN             NaN   NaN   
Service                        3493.0    NaN             NaN   NaN   
Comfortablility                3493.0    NaN             NaN   NaN   
Privacy                        3493.0    NaN             NaN   NaN   

                                          mean              std        min  \
Registration Number                    61747.0      1008.486572    60001.0   
Annual Turnover                30725708.559977  21651246.901146  3500000.0   
Cuisine                                    NaN              NaN        NaN   
City                                       NaN              NaN        NaN   
Restaurant Location                        NaN              NaN        NaN   
Opening Day of Restaurant                  NaN              NaN        NaN   
Facebook Popularity Quotient         77.938715         9.829169       43.0   
Endorsed By                                NaN              NaN        NaN   
Instagram Popularity Quotient        74.404684        10.940327       40.0   
Fire Audit                             0.78872         0.408275        0.0   
Liquor License Obtained               0.988262         0.107719        0.0   
Situated in a Multi Complex           0.808188         0.393783        0.0   
Dedicated Parking                     0.801889         0.398633        0.0   
Open Sitting Available                0.800172         0.399928        0.0   
Resturant Tier                        1.926539          0.26093        1.0   
Restaurant Type                            NaN              NaN        NaN   
Restaurant Theme                           NaN              NaN        NaN   
Restaurant Zomato Rating              2.696536         0.787271        0.0   
Restaurant City Tier                   0.30146         0.458958        0.0   
Order Wait Time      

In [6]:
data.isnull().sum()

Registration Number                 0
Annual Turnover                     0
Cuisine                             0
City                                0
Restaurant Location                 0
Opening Day of Restaurant           0
Facebook Popularity Quotient       99
Endorsed By                         0
Instagram Popularity Quotient      56
Fire Audit                          0
Liquor License Obtained             0
Situated in a Multi Complex         0
Dedicated Parking                   0
Open Sitting Available              0
Resturant Tier                     49
Restaurant Type                     0
Restaurant Theme                    0
Restaurant Zomato Rating            0
Restaurant City Tier                0
Order Wait Time                     0
Staff Responsivness                 0
Value for Money                     0
Hygiene Rating                      0
Food Rating                         0
Overall Restaurant Rating         212
Live Music Rating                 765
Comedy Gigs 

In [7]:
data['City'].value_counts()

City
Bangalore    553
-1           396
Noida        324
Hyderabad    295
Pune         262
            ... 
bihar          1
singaruli      1
KANPUR         1
Banglore       1
pondy          1
Name: count, Length: 297, dtype: int64

#### Data Overview Initial Findings:

We have a few issues at play that we'll need to take care of:

* There are 297 unique cities and 396 of the 3,493 entries have an incorrect value as `-1` is definitely not a valid city
  * Instead of using the actual city, we can try and generalize it more to a larger area

* The null values for `Live Music Rating`, `Comedy Gigs Rating`, `Value Deals Rating`, and `Live Sports Rating` aren't actually null, but represent places that do not offer these services.  Filling in those missing values with a value of `0` would make more sense.  However, it could also be that creating a new feature like `Has Live Music` may be more informative.

While the above does require some data imputation this shouldn't result in any data spillage between the training and validation sets that will be used later on.

For the remaining null values we can use mice forest imputation after on our training data set and apply that to our validation set, test set, and finally the final set that will be used for the submission.  This should ensure we avoid any data spillage between the data sets while still allowing us to do an in-depth exploratory data analysis.

### Ratings Imputation and Ratings Feature Creation

In [8]:
# Function to impute ratings 
def impute_ratings(dataset):
    
    dataset['Live Music Rating'] = dataset['Live Music Rating'].fillna(0)
    dataset['Comedy Gigs Rating'] = dataset['Comedy Gigs Rating'].fillna(0)
    dataset['Value Deals Rating'] = dataset['Value Deals Rating'].fillna(0)
    dataset['Live Sports Rating'] = dataset['Live Sports Rating'].fillna(0)
    
    return dataset

# Function to create new features for ratings
def create_ratings_feature(dataset):
    
    dataset['Has_Live_Music'] = dataset['Live Music Rating'].apply(lambda x: 1 if x > 0 else 0)
    dataset['Has_Comedy_Gigs'] = dataset['Comedy Gigs Rating'].apply(lambda x: 1 if x > 0 else 0)
    dataset['Has_Value_Deals'] = dataset['Value Deals Rating'].apply(lambda x: 1 if x > 0 else 0)
    dataset['Has_Live_Sports'] = dataset['Live Sports Rating'].apply(lambda x: 1 if x > 0 else 0)
    
    return dataset

In [9]:
data = impute_ratings(data)
data.isnull().sum()

Registration Number                0
Annual Turnover                    0
Cuisine                            0
City                               0
Restaurant Location                0
Opening Day of Restaurant          0
Facebook Popularity Quotient      99
Endorsed By                        0
Instagram Popularity Quotient     56
Fire Audit                         0
Liquor License Obtained            0
Situated in a Multi Complex        0
Dedicated Parking                  0
Open Sitting Available             0
Resturant Tier                    49
Restaurant Type                    0
Restaurant Theme                   0
Restaurant Zomato Rating           0
Restaurant City Tier               0
Order Wait Time                    0
Staff Responsivness                0
Value for Money                    0
Hygiene Rating                     0
Food Rating                        0
Overall Restaurant Rating        212
Live Music Rating                  0
Comedy Gigs Rating                 0
V

In [10]:
data = create_ratings_feature(data)
data.head()

Registration Number  Annual Turnover           Cuisine       City  \
0                60001         42000000      indian,irish  Bangalore   
1                60002         50000000      indian,irish     Indore   
2                60003         32500000   tibetan,italian    Chennai   
3                60004        110000000  turkish,nigerian    Gurgaon   
4                60005         20000000     irish,belgian    Manesar   

  Restaurant Location Opening Day of Restaurant  Facebook Popularity Quotient  \
0   Near Business Hub                14-02-2009                          84.3   
1      Near Party Hub                29-09-2008                          85.4   
2   Near Business Hub                30-07-2011                          85.0   
3      Near Party Hub                30-11-2008                          85.6   
4      Near Party Hub                22-02-2010                           NaN   

        Endorsed By  Instagram Popularity Quotient  Fire Audit  ...  \
0      Not Specific                           95.8           1  ...   
1  Tier A Celebrity                           85.0           1  ...   
2  Tier A Celebrity                           68.2           1  ...   
3  Tier A Celebrity                           83.6           0  ...   
4  Tier A Celebrity                           76.8           1  ...   

   Live Sports Rating  Ambience  Lively  Service  Comfortablility Privacy  \
0                 0.0       8.0       8        6                6       6   
1                 0.0       5.0       7        7                3       8   
2                 0.0       7.0      10        5                2       8   
3                 0.0       7.0       7        4                3       5   
4                 0.0       NaN       6        2                4       6   

  Has_Live_Music  Has_Comedy_Gigs  Has_Value_Deals  Has_Live_Sports  
0              1                0                0                0  
1              0                1                0                0  
2              1                0                0                0  
3              1                0                0                0  
4              0                1                0                0  

[5 rows x 38 columns]

We have a lot less null values now and we have a few new features that may help us out in our analysis

### City Matching and Generalization

In [11]:
# Load in our geographical data for cities and towns
# Loading data for hamlets and villages resulted in our fuzzy search taking WAY too long
cities = pd.read_json('data/place-city.ndjson', lines=True)
towns = pd.read_json('data/place-town.ndjson', lines=True)

all_locations = pd.concat([cities, towns], axis=0)
all_locations.head()

name                                        other_names  \
0  Puthanathani  {'name:hi': 'पुतनतानी', 'name:ml': 'പുത്തനത്താ...   
1    Malappuram  {'name:ar': 'مالابورام', 'name:en': 'Malappura...   
2           NaN                                                NaN   
3        Kannur  {'name:mr': 'कण्णूर', 'name:ar': 'كانور', 'nam...   
4         Ponda  {'name:en': 'Ponda', 'name:hi': 'पोण्डा', 'nam...   

                                        display_name  \
0  Puthanathani, Tirur, Malappuram, Kerala, 67655...   
1       Malappuram, Ernad, Malappuram, Kerala, India   
2  Tiruchchirāppalli, Tiruchirappalli District, T...   
3             Kannur, Kannur district, Kerala, India   
4               Ponda, South Goa, Goa, 403401, India   

                                             address  population  osm_type  \
0  {'city': 'Puthanathani', 'county': 'Tirur', 's...     20480.0       way   
1  {'town': 'Malappuram', 'county': 'Ernad', 'sta...    101330.0       way   
2  {'county': 'Tiruchchirāppalli', 'state_distric...         NaN       way   
3  {'city': 'Kannur', 'county': 'Kannur', 'state_...     56823.0       way   
4  {'city': 'Ponda', 'county': 'Ponda', 'state_di...         NaN  relation   

      osm_id            type                                 location  \
0  783681659            city         [76.00462108208228, 10.93790995]   
1   84635269            city                 [76.0807838, 11.0428925]   
2   84652860            city  [78.70113360377684, 10.791470050000001]   
3   22825987            city                 [75.3737973, 11.8763836]   
4    8614127  administrative  [74.00797876754456, 15.397935050000001]   

                                               bbox  
0   [75.9927547, 10.926375, 76.0160868, 10.9477521]  
1  [76.0338675, 11.0263125, 76.1033108, 11.0964078]  
2  [78.6207515, 10.7398647, 78.7467839, 10.8401688]  
3  [75.3528857, 11.8516327, 75.4042578, 11.9237909]  
4  [73.9823804, 15.3763776, 74.0399122, 15.4189111]

In [12]:
# We really only need the name, other_names, and address columns which we'll explode to make create a new column for state
locations_subset = all_locations[['name', 'other_names', 'address']]
locations_subset['state'] = pd.json_normalize(locations_subset['address'])['state']
locations_subset.head()

name                                        other_names  \
0  Puthanathani  {'name:hi': 'पुतनतानी', 'name:ml': 'പുത്തനത്താ...   
1    Malappuram  {'name:ar': 'مالابورام', 'name:en': 'Malappura...   
2           NaN                                                NaN   
3        Kannur  {'name:mr': 'कण्णूर', 'name:ar': 'كانور', 'nam...   
4         Ponda  {'name:en': 'Ponda', 'name:hi': 'पोण्डा', 'nam...   

                                             address       state  
0  {'city': 'Puthanathani', 'county': 'Tirur', 's...      Kerala  
1  {'town': 'Malappuram', 'county': 'Ernad', 'sta...      Kerala  
2  {'county': 'Tiruchchirāppalli', 'state_distric...  Tamil Nadu  
3  {'city': 'Kannur', 'county': 'Kannur', 'state_...      Kerala  
4  {'city': 'Ponda', 'county': 'Ponda', 'state_di...         Goa

In [13]:
# Now we'll explode the other_names column to grab a few variations where they make sense
locations_subset['alt_name'] = pd.json_normalize(locations_subset['other_names'])['alt_name']
locations_subset['old_name'] = pd.json_normalize(locations_subset['other_names'])['old_name']
locations_subset['english_name'] = pd.json_normalize(locations_subset['other_names'])['name:en']
locations_subset.head()

name                                        other_names  \
0  Puthanathani  {'name:hi': 'पुतनतानी', 'name:ml': 'പുത്തനത്താ...   
1    Malappuram  {'name:ar': 'مالابورام', 'name:en': 'Malappura...   
2           NaN                                                NaN   
3        Kannur  {'name:mr': 'कण्णूर', 'name:ar': 'كانور', 'nam...   
4         Ponda  {'name:en': 'Ponda', 'name:hi': 'पोण्डा', 'nam...   

                                             address       state alt_name  \
0  {'city': 'Puthanathani', 'county': 'Tirur', 's...      Kerala      NaN   
1  {'town': 'Malappuram', 'county': 'Ernad', 'sta...      Kerala      NaN   
2  {'county': 'Tiruchchirāppalli', 'state_distric...  Tamil Nadu      NaN   
3  {'city': 'Kannur', 'county': 'Kannur', 'state_...      Kerala      NaN   
4  {'city': 'Ponda', 'county': 'Ponda', 'state_di...         Goa      NaN   

    old_name english_name  
0        NaN          NaN  
1        NaN   Malappuram  
2        NaN          NaN  
3  Cannanore       Kannur  
4        NaN        Ponda

In [14]:
locations_subset[locations_subset['state'].isnull()]

name                                        other_names  \
70   Chandigarh                        {'name:ur': 'ضلع چنڈی گڑھ'}   
427  Chandigarh  {'name:ar': 'شانديغار', 'name:cs': 'Čandígarh'...   
70    Attamaala                                                NaN   
427    Mukoodal    {'name:ml': 'മുക്കൂടൽ', 'name:ta': 'முக்கூடல்'}   

                                               address state alt_name  \
70   {'city': 'Chandigarh', 'city_district': 'Chand...   NaN      NaN   
427  {'city': 'Chandigarh', 'ISO3166-2-lvl4': 'IN-C...   NaN   चंडीगढ़   
70   {'town': 'Attamaala', 'state_district': 'Wayan...   NaN      NaN   
427  {'town': 'Mukoodal', 'county': 'Cheranmahadevi...   NaN   चंडीगढ़   

    old_name english_name  
70       NaN          NaN  
427      NaN   Chandigarh  
70       NaN          NaN  
427      NaN   Chandigarh

For the most part it looks like the states that are null are from villages and towns and may be due to them being such a small size.  However, the city of `Chandigarh` seems pretty important and after doing some research it appears it serves as a [capital of two states](https://en.wikipedia.org/wiki/Politics_of_Chandigarh#:~:text=Chandigarh%20is%20a%20city%20and,not%20part%20of%20either%20state.) so it may make sense to define this as a region on its own.

In [15]:
locations_subset['state'][locations_subset['name'] == 'Chandigarh'] = 'Chandigarh'
locations_subset[locations_subset['state'].isnull()]

name                                      other_names  \
70   Attamaala                                              NaN   
427   Mukoodal  {'name:ml': 'മുക്കൂടൽ', 'name:ta': 'முக்கூடல்'}   

                                               address state alt_name  \
70   {'town': 'Attamaala', 'state_district': 'Wayan...   NaN      NaN   
427  {'town': 'Mukoodal', 'county': 'Cheranmahadevi...   NaN   चंडीगढ़   

    old_name english_name  
70       NaN          NaN  
427      NaN   Chandigarh

In [16]:
# Drop the rest of the columns where the state column is null and create a set for all geographical names to be used in our fuzzy search
#locations_subset.drop(locations_subset[locations_subset['state'].isnull()], axis=1, inplace=True)
locations_subset = locations_subset.dropna(subset=['state'], axis=0)
locations_subset[locations_subset['state'].isnull()]


Empty DataFrame
Columns: [name, other_names, address, state, alt_name, old_name, english_name]
Index: []

In [17]:
locations_subset.head()

name                                        other_names  \
0  Puthanathani  {'name:hi': 'पुतनतानी', 'name:ml': 'പുത്തനത്താ...   
1    Malappuram  {'name:ar': 'مالابورام', 'name:en': 'Malappura...   
2           NaN                                                NaN   
3        Kannur  {'name:mr': 'कण्णूर', 'name:ar': 'كانور', 'nam...   
4         Ponda  {'name:en': 'Ponda', 'name:hi': 'पोण्डा', 'nam...   

                                             address       state alt_name  \
0  {'city': 'Puthanathani', 'county': 'Tirur', 's...      Kerala      NaN   
1  {'town': 'Malappuram', 'county': 'Ernad', 'sta...      Kerala      NaN   
2  {'county': 'Tiruchchirāppalli', 'state_distric...  Tamil Nadu      NaN   
3  {'city': 'Kannur', 'county': 'Kannur', 'state_...      Kerala      NaN   
4  {'city': 'Ponda', 'county': 'Ponda', 'state_di...         Goa      NaN   

    old_name english_name  
0        NaN          NaN  
1        NaN   Malappuram  
2        NaN          NaN  
3  Cannanore       Kannur  
4        NaN        Ponda

In [18]:
locations_subset.shape

(4843, 7)

In [19]:
# Building a set with all geographical names to use for our fuzzy search
search_locations = set()

for index, row in locations_subset.iterrows():
    search_locations.add(str(row['name']).lower())
    search_locations.add(str(row['alt_name']).lower())
    search_locations.add(str(row['old_name']).lower())
    search_locations.add(str(row['english_name']).lower())
    
print(f"There are {len(search_locations)} searchable locations")

There are 4795 searchable locations


In [20]:
data['City'] = data['City'].replace('-1', np.nan)

In [21]:
# Testing out our fuzzy search to see how many matches are found
# Commenting out to prevent running on rerun
# for index, row in data.iterrows():
    # current_location = str(row['City']).lower()
    # found = process.extractOne(current_location, search_locations)
    # if(found[1] < 95):
        # print(f"{current_location} closest match was {found[0]} with confidence of {found[1]}%")
        

For the most part, it looks like anything with a confidence interval of higher than 85% does a good job at matching correctly.  The only exceptions found would be the following entries which should be double checked:

* australia closest match was tral with confidence of 90%
  * Not in India
* indirapuram, ghaziabad closest match was indi with confidence of 90%
  * Should be changed to Ghaziabad
* meerut closest match was eru with confidence of 90%
  * Located in Uttar Pradesh
* ncr closest match was puthencruz with confidence of 90%
  * Another entry for delhi/ncr was found so this should probably be changed to that
* am closest match was amritsar with confidence of 90%
  * This one is far too generic to attempt a fuzzy search on so we may just null it out and let our imputation handle it at a later state
* tornagallu closest match was nagal with confidence of 90%
  * Village in Karnataka (Will manually map)
* trichur closest match was raichur with confidence of 86%
  * Google auto corrects this to Thrissur which looks like a better match than what fuzzy search found; will manually map to Kerala
* ras al khaimah closest match was mahé with confidence of 90%
  * Not in India (part of UAE)
* singaruli closest match was singa with confidence of 90%
  * Looks like this should be singrauli which is in Madhya Pradesh

Additionally, there were entries for `London` and `Dubai`.  London is definitely not in India and Dubai is also a part of the United Arab Emirates.  Let's take a look at how many entries we have for those cities

In [22]:
data[(data['City'] == 'LONDON') | (data['City'] == 'Australia') | (data['City'] == 'Dubai') | (data['City'] == 'RAS AL KHAIMAH') | (data['City'] == 'AM')]

Registration Number  Annual Turnover              Cuisine  \
669                 60670         48000000        cuban,british   
927                 60928        130000000        polish,jewish   
1439                61440         36000000        tibetan,greek   
1746                61747        132000000     hawaiian,latvian   
2536                62537        200000000      algerian,korean   
2970                62971         80000000        irish,belgian   
3437                63438         40000000  chinese,salvadorian   

                City Restaurant Location Opening Day of Restaurant  \
669        Australia      Near Party Hub                16-10-2011   
927            Dubai      Near Party Hub                24-07-2009   
1439              AM      Near Party Hub                07-12-2007   
1746           Dubai      Near Party Hub                11-09-2006   
2536          LONDON      Near Party Hub                13-03-2012   
2970           Dubai      Near Party Hub                17-03-2009   
3437  RAS AL KHAIMAH      Near Party Hub                20-02-2009   

      Facebook Popularity Quotient       Endorsed By  \
669                          64.30      Not Specific   
927                          80.00      Not Specific   
1439                         73.06      Not Specific   
1746                           NaN      Not Specific   
2536                         90.00  Tier A Celebrity   
2970                         83.40  Tier A Celebrity   
3437                         87.40      Not Specific   

      Instagram Popularity Quotient  Fire Audit  ...  Live Sports Rating  \
669                            66.4           1  ...                 0.0   
927                            90.0           1  ...                 0.0   
1439                           74.0           1  ...                 4.0   
1746                           72.0           1  ...                 0.0   
2536                           89.4           1  ...                 0.0   
2970                           70.0           1  ...                 0.0   
3437                           76.0           1  ...                 0.0   

      Ambience  Lively  Service  Comfortablility Privacy Has_Live_Music  \
669       10.0       3        2                0       4              1   
927       10.0      10        6                5       7              1   
1439       1.0       2        3                4       3              1   
1746       4.0       9        5                4       7              1   
2536       9.0      10        5                0       6              1   
2970       7.0       8        4                3       8              1   
3437       4.0       4        5                0       3              1   

      Has_Comedy_Gigs  Has_Value_Deals  Has_Live_Sports  
669                 0                1                0  
927                 0                0                0  
1439                1                0                1  
1746                1                0                0  
2536                0                1                0  
2970                0                1                0  
3437                1                0                0  

[7 rows x 38 columns]

In [23]:
# We will go ahead and drop the rows as there are only 6 which should have no impact on our dataset
# If anything it should help our predictions as these entries being from other countries would certainly skew our results
def drop_invalid_cities(dataset):
    
    dataset.drop(dataset[dataset['City'] == 'Dubai'].index, inplace=True)
    dataset.drop(dataset[dataset['City'] == 'LONDON'].index, inplace=True)
    dataset.drop(dataset[dataset['City'] == 'Australia'].index, inplace=True)
    dataset.drop(dataset[dataset['City'] == 'RAS AL KHAIMAH'].index, inplace=True)
    dataset.drop(dataset[dataset['City'] == 'AM'].index, inplace=True)
    
    return dataset

data = drop_invalid_cities(data)

# Should now have 3486 rows (3493 - 7)
data.shape

(3486, 38)

In [24]:
mismatched_locations = {
    "indirapuram, ghaziabad": "Uttar Pradesh",
    "ncr": "Delhi",
    "singaruli": "Madhya Pradesh",
    "meerut": "Uttar Pradesh",
    "tornagallu": "Karnataka",
    "trichur": "Kerala",
    "singaruli": "Madhya Pradesh"
}

In [25]:
# Set to hold locations that were not found
# Commenting out to prevent running on rerun
# locations_not_found = set()

# for index, row in data.iterrows():
#     current_location = str(row['City']).lower()
#     # Use our manual mapping first
#     if (current_location in mismatched_locations.keys()):
#         print(f'Manually mapping to {mismatched_locations[current_location]}')
#     # Use Fuzzy search if no manual mappings are found
#     else:
#         found = process.extractOne(current_location, search_locations)
#         if(found[1] < 85):
#             locations_not_found.add(current_location)
#             print(f"{current_location} closest match was {found[0]} with confidence of {found[1]}%")

# locations_not_found

In [26]:
# Added in mappings from mismatched dictionary to this one
manual_mappings = {
    "indirapuram, ghaziabad": "Uttar Pradesh",
    "ncr": "Delhi",
    "singaruli": "Madhya Pradesh",
    "meerut": "Uttar Pradesh",
    "tornagallu": "Karnataka",
    "trichur": "Kerala",
    "singaruli": "Madhya Pradesh",
    "gagret": "Himachal Pradesh",
    "gajiabaad": "Uttar Pradesh",
    "gurga": "Haryana",
    "keral": "Kerala",
    "nasikcity": "Maharashtra",
    "nouda": "Uttar Pradesh",
    "orissa": "Odisha",
    "pondy": "Puducherry",
    "punr": "Maharashtra",
    "rajasthan": "Rajasthan",
    "sahibabad": "Uttar Pradesh",
    "trichy": "Tamil Nadu"
}

# Create reusable function to map cities to states
def map_cities_to_state(dataset):
    for index, row in dataset.iterrows():
        current_location = str(row['City']).lower()
        # Only do search if city is not null
        if (current_location != 'nan'):
            # Use our manual mapping first
            if (current_location in manual_mappings.keys()):
                print(f'Manually mapping {current_location} to {manual_mappings[current_location]}')
                dataset.loc[index, 'state'] = manual_mappings[current_location]
            # Use Fuzzy search if no manual mappings are found
            else:
                found = process.extractOne(current_location, search_locations)
                found_state = locations_subset[
                    (locations_subset['name'].str.lower() == found[0]) | 
                    (locations_subset['old_name'].str.lower() == found[0]) | 
                    (locations_subset['alt_name'].str.lower() == found[0]) | 
                    (locations_subset['english_name'].str.lower() == found[0])
                ]['state'].head(1).values[0]
                dataset.loc[index, 'state'] = found_state
    return dataset

data = map_cities_to_state(data)
data.head()


Manually mapping nouda to Uttar Pradesh
Manually mapping gurga to Haryana
Manually mapping gagret to Himachal Pradesh
Manually mapping indirapuram, ghaziabad to Uttar Pradesh
Manually mapping rajasthan to Rajasthan
Manually mapping meerut to Uttar Pradesh
Manually mapping ncr to Delhi
Manually mapping trichy to Tamil Nadu
Manually mapping trichy to Tamil Nadu
Manually mapping nasikcity to Maharashtra
Manually mapping tornagallu to Karnataka
Manually mapping gajiabaad to Uttar Pradesh
Manually mapping trichy to Tamil Nadu
Manually mapping trichy to Tamil Nadu
Manually mapping trichy to Tamil Nadu
Manually mapping nasikcity to Maharashtra
Manually mapping orissa to Odisha
Manually mapping nasikcity to Maharashtra
Manually mapping punr to Maharashtra
Manually mapping meerut to Uttar Pradesh
Manually mapping sahibabad to Uttar Pradesh
Manually mapping trichur to Kerala
Manually mapping keral to Kerala
Manually mapping nasikcity to Maharashtra
Manually mapping singaruli to Madhya Pradesh
Ma

Registration Number  Annual Turnover           Cuisine       City  \
0                60001         42000000      indian,irish  Bangalore   
1                60002         50000000      indian,irish     Indore   
2                60003         32500000   tibetan,italian    Chennai   
3                60004        110000000  turkish,nigerian    Gurgaon   
4                60005         20000000     irish,belgian    Manesar   

  Restaurant Location Opening Day of Restaurant  Facebook Popularity Quotient  \
0   Near Business Hub                14-02-2009                          84.3   
1      Near Party Hub                29-09-2008                          85.4   
2   Near Business Hub                30-07-2011                          85.0   
3      Near Party Hub                30-11-2008                          85.6   
4      Near Party Hub                22-02-2010                           NaN   

        Endorsed By  Instagram Popularity Quotient  Fire Audit  ...  Ambience  \
0      Not Specific                           95.8           1  ...       8.0   
1  Tier A Celebrity                           85.0           1  ...       5.0   
2  Tier A Celebrity                           68.2           1  ...       7.0   
3  Tier A Celebrity                           83.6           0  ...       7.0   
4  Tier A Celebrity                           76.8           1  ...       NaN   

   Lively  Service  Comfortablility  Privacy Has_Live_Music Has_Comedy_Gigs  \
0       8        6                6        6              1               0   
1       7        7                3        8              0               1   
2      10        5                2        8              1               0   
3       7        4                3        5              1               0   
4       6        2                4        6              0               1   

   Has_Value_Deals  Has_Live_Sports           state  
0                0                0       Karnataka  
1                0                0  Madhya Pradesh  
2                0                0      Tamil Nadu  
3                0                0         Haryana  
4                0                0         Haryana  

[5 rows x 39 columns]

In [27]:
data.isnull().sum()

Registration Number                0
Annual Turnover                    0
Cuisine                            0
City                             396
Restaurant Location                0
Opening Day of Restaurant          0
Facebook Popularity Quotient      98
Endorsed By                        0
Instagram Popularity Quotient     56
Fire Audit                         0
Liquor License Obtained            0
Situated in a Multi Complex        0
Dedicated Parking                  0
Open Sitting Available             0
Resturant Tier                    49
Restaurant Type                    0
Restaurant Theme                   0
Restaurant Zomato Rating           0
Restaurant City Tier               0
Order Wait Time                    0
Staff Responsivness                0
Value for Money                    0
Hygiene Rating                     0
Food Rating                        0
Overall Restaurant Rating        212
Live Music Rating                  0
Comedy Gigs Rating                 0
V

In [28]:
data.describe(include='all').T

count unique             top  freq  \
Registration Number            3486.0    NaN             NaN   NaN   
Annual Turnover                3486.0    NaN             NaN   NaN   
Cuisine                          3486     20   tibetan,greek   685   
City                             3090    291       Bangalore   553   
Restaurant Location              3486      2  Near Party Hub  2651   
Opening Day of Restaurant        3486   1776      27-12-2009    10   
Facebook Popularity Quotient   3388.0    NaN             NaN   NaN   
Endorsed By                      3486      3    Not Specific  1956   
Instagram Popularity Quotient  3430.0    NaN             NaN   NaN   
Fire Audit                     3486.0    NaN             NaN   NaN   
Liquor License Obtained        3486.0    NaN             NaN   NaN   
Situated in a Multi Complex    3486.0    NaN             NaN   NaN   
Dedicated Parking              3486.0    NaN             NaN   NaN   
Open Sitting Available         3486.0    NaN             NaN   NaN   
Resturant Tier                 3437.0    NaN             NaN   NaN   
Restaurant Type                  3486      4             Bar  3229   
Restaurant Theme                 3486     34         Arabian  1369   
Restaurant Zomato Rating       3486.0    NaN             NaN   NaN   
Restaurant City Tier           3486.0    NaN             NaN   NaN   
Order Wait Time                3486.0    NaN             NaN   NaN   
Staff Responsivness            3486.0    NaN             NaN   NaN   
Value for Money                3486.0    NaN             NaN   NaN   
Hygiene Rating                 3486.0    NaN             NaN   NaN   
Food Rating                    3486.0    NaN             NaN   NaN   
Overall Restaurant Rating      3274.0    NaN             NaN   NaN   
Live Music Rating              3486.0    NaN             NaN   NaN   
Comedy Gigs Rating             3486.0    NaN             NaN   NaN   
Value Deals Rating             3486.0    NaN             NaN   NaN   
Live Sports Rating             3486.0    NaN             NaN   NaN   
Ambience                       3461.0    NaN             NaN   NaN   
Lively                         3486.0    NaN             NaN   NaN   
Service                        3486.0    NaN             NaN   NaN   
Comfortablility                3486.0    NaN             NaN   NaN   
Privacy                        3486.0    NaN             NaN   NaN   
Has_Live_Music                 3486.0    NaN             NaN   NaN   
Has_Comedy_Gigs                3486.0    NaN             NaN   NaN   
Has_Value_Deals                3486.0    NaN             NaN   NaN   
Has_Live_Sports                3486.0    NaN             NaN   NaN   
state                            3090     23       Karnataka   657   

                                          mean              std        min  \
Registration Number               61746.569134      1008.516728    60001.0   
Annual Turnover                30596356.855995  21328229.074323  3500000.0   
Cuisine                                    NaN              NaN        NaN   
City                                       NaN              NaN        NaN   
Restaurant Location                        NaN              NaN        NaN   
Opening Day of Restaurant                  NaN              NaN        NaN   
Facebook Popularity Quotient         77.935608         9.830682       43.0   
Endorsed By                                NaN              NaN        NaN   
Instagram Popularity Quotient        74.399738        10.944028       40.0   
Fire Audit                            0.788296         0.408575        0.0   
Liquor License Obtained               0.988239         0.107825        0.0   
Situated in a Multi Complex            0.80809          0.39386        0.0   
Dedicated Parking                     0.801779         0.398717        0.0   
Open Sitting Available                0.800057         0.400014        0.0   
Resturant Tier                         1.92668         0.2606

We've gone from 291 cities to 23 different states which is much more manageable.  The last part that bugs me a bit about this data is how cuisines have been paired together as I would think specific cuisine types might be more expensive so let's split that out and create new features from it.  Additionally, we can't do much with the datetime value in `Opening Day of Restaurant` so let's convert that to something more like `Years in Business`.

### Final Data Cleanup and Checks

In [29]:
def convert_time(dataset):
    
    dataset['Opening Day of Restaurant'] = pd.to_datetime(dataset['Opening Day of Restaurant'])
    dataset['Years in Business'] = 2025 - dataset['Opening Day of Restaurant'].dt.year
    dataset.drop('Opening Day of Restaurant', inplace=True, axis=1)
    
    return dataset

data = convert_time(data)
data.head()

Registration Number  Annual Turnover           Cuisine       City  \
0                60001         42000000      indian,irish  Bangalore   
1                60002         50000000      indian,irish     Indore   
2                60003         32500000   tibetan,italian    Chennai   
3                60004        110000000  turkish,nigerian    Gurgaon   
4                60005         20000000     irish,belgian    Manesar   

  Restaurant Location  Facebook Popularity Quotient       Endorsed By  \
0   Near Business Hub                          84.3      Not Specific   
1      Near Party Hub                          85.4  Tier A Celebrity   
2   Near Business Hub                          85.0  Tier A Celebrity   
3      Near Party Hub                          85.6  Tier A Celebrity   
4      Near Party Hub                           NaN  Tier A Celebrity   

   Instagram Popularity Quotient  Fire Audit  Liquor License Obtained  ...  \
0                           95.8           1                        1  ...   
1                           85.0           1                        1  ...   
2                           68.2           1                        1  ...   
3                           83.6           0                        1  ...   
4                           76.8           1                        1  ...   

   Lively  Service  Comfortablility  Privacy Has_Live_Music Has_Comedy_Gigs  \
0       8        6                6        6              1               0   
1       7        7                3        8              0               1   
2      10        5                2        8              1               0   
3       7        4                3        5              1               0   
4       6        2                4        6              0               1   

   Has_Value_Deals  Has_Live_Sports           state  Years in Business  
0                0                0       Karnataka                 16  
1                0                0  Madhya Pradesh                 17  
2                0                0      Tamil Nadu                 14  
3                0                0         Haryana                 17  
4                0                0         Haryana                 15  

[5 rows x 39 columns]

In [40]:
def convert_restaurant_location(dataset):
    
    dataset['Located_Near_Party_Hub'] = dataset['Restaurant Location'].apply(lambda x: 1 if x == 'Near Party Hub' else 0)
    dataset.drop('Restaurant Location', inplace=True, axis=1)
    
    return dataset

In [31]:
def explode_cuisine(dataset):
    
    cuisine_split = dataset['Cuisine'].str.get_dummies(sep=',')
    merged_dataset = pd.concat([dataset, cuisine_split.add_prefix('Has_')], axis=1)

    return merged_dataset

In [46]:
data = convert_restaurant_location(data)
data = explode_cuisine(data)
data.head()

Registration Number  Annual Turnover           Cuisine       City  \
0                60001         42000000      indian,irish  Bangalore   
1                60002         50000000      indian,irish     Indore   
2                60003         32500000   tibetan,italian    Chennai   
3                60004        110000000  turkish,nigerian    Gurgaon   
4                60005         20000000     irish,belgian    Manesar   

   Facebook Popularity Quotient       Endorsed By  \
0                          84.3      Not Specific   
1                          85.4  Tier A Celebrity   
2                          85.0  Tier A Celebrity   
3                          85.6  Tier A Celebrity   
4                           NaN  Tier A Celebrity   

   Instagram Popularity Quotient  Fire Audit  Liquor License Obtained  \
0                           95.8           1                        1   
1                           85.0           1                        1   
2                           68.2           1                        1   
3                           83.6           0                        1   
4                           76.8           1                        1   

   Situated in a Multi Complex  ...  Has_nigerian  Has_peruvian  Has_polish  \
0                            1  ...             0             0           0   
1                            1  ...             0             0           0   
2                            1  ...             0             0           0   
3                            1  ...             1             0           0   
4                            1  ...             0             0           0   

  Has_salvadorian Has_sapnish  Has_swedish  Has_thai  Has_tibetan  \
0               0           0            0         0            0   
1               0           0            0         0            0   
2               0           0            0         0            1   
3               0           0            0         0            0   
4               0           0            0         0            0   

   Has_turkish  Has_welsh  
0            0          0  
1            0          0  
2            0          0  
3            1          0  
4            0          0  

[5 rows x 64 columns]

In [47]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3486 entries, 0 to 3492
Data columns (total 64 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Registration Number            3486 non-null   int64  
 1   Annual Turnover                3486 non-null   int64  
 2   Cuisine                        3486 non-null   object 
 3   City                           3090 non-null   object 
 4   Facebook Popularity Quotient   3388 non-null   float64
 5   Endorsed By                    3486 non-null   object 
 6   Instagram Popularity Quotient  3430 non-null   float64
 7   Fire Audit                     3486 non-null   int64  
 8   Liquor License Obtained        3486 non-null   int64  
 9   Situated in a Multi Complex    3486 non-null   int64  
 10  Dedicated Parking              3486 non-null   int64  
 11  Open Sitting Available         3486 non-null   int64  
 12  Resturant Tier                 3437 non-null   float6

In [52]:
def convert_objects_to_categorical(dataset):
    
    object_columns = dataset.select_dtypes('object').columns
    dataset[object_columns] = dataset[object_columns].astype('category')
    
    return dataset

data = convert_objects_to_categorical(data)
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3486 entries, 0 to 3492
Data columns (total 64 columns):
 #   Column                         Non-Null Count  Dtype   
---  ------                         --------------  -----   
 0   Registration Number            3486 non-null   int64   
 1   Annual Turnover                3486 non-null   int64   
 2   Cuisine                        3486 non-null   category
 3   City                           3090 non-null   category
 4   Facebook Popularity Quotient   3388 non-null   float64 
 5   Endorsed By                    3486 non-null   category
 6   Instagram Popularity Quotient  3430 non-null   float64 
 7   Fire Audit                     3486 non-null   int64   
 8   Liquor License Obtained        3486 non-null   int64   
 9   Situated in a Multi Complex    3486 non-null   int64   
 10  Dedicated Parking              3486 non-null   int64   
 11  Open Sitting Available         3486 non-null   int64   
 12  Resturant Tier                 3437 non

In [53]:
# Check for duplicates
data.duplicated().sum()

np.int64(0)

In [54]:
# Drop columns that will provide no value
data.drop('Registration Number', axis=1, inplace=True)
data.head()

Annual Turnover           Cuisine       City  Facebook Popularity Quotient  \
0         42000000      indian,irish  Bangalore                          84.3   
1         50000000      indian,irish     Indore                          85.4   
2         32500000   tibetan,italian    Chennai                          85.0   
3        110000000  turkish,nigerian    Gurgaon                          85.6   
4         20000000     irish,belgian    Manesar                           NaN   

        Endorsed By  Instagram Popularity Quotient  Fire Audit  \
0      Not Specific                           95.8           1   
1  Tier A Celebrity                           85.0           1   
2  Tier A Celebrity                           68.2           1   
3  Tier A Celebrity                           83.6           0   
4  Tier A Celebrity                           76.8           1   

   Liquor License Obtained  Situated in a Multi Complex  Dedicated Parking  \
0                        1                            1                  0   
1                        1                            1                  1   
2                        1                            1                  1   
3                        1                            1                  1   
4                        1                            1                  1   

   ...  Has_nigerian  Has_peruvian Has_polish Has_salvadorian  Has_sapnish  \
0  ...             0             0          0               0            0   
1  ...             0             0          0               0            0   
2  ...             0             0          0               0            0   
3  ...             1             0          0               0            0   
4  ...             0             0          0               0            0   

   Has_swedish  Has_thai  Has_tibetan  Has_turkish  Has_welsh  
0            0         0            0            0          0  
1            0         0            0            0          0  
2            0         0            1            0          0  
3            0         0            0            1          0  
4            0         0            0            0          0  

[5 rows x 63 columns]

## Exploratory Data Analysis

In [55]:
# Creating copy of data as we may tweak it a bit during analysis
eda = data.copy()
eda.head()

Annual Turnover           Cuisine       City  Facebook Popularity Quotient  \
0         42000000      indian,irish  Bangalore                          84.3   
1         50000000      indian,irish     Indore                          85.4   
2         32500000   tibetan,italian    Chennai                          85.0   
3        110000000  turkish,nigerian    Gurgaon                          85.6   
4         20000000     irish,belgian    Manesar                           NaN   

        Endorsed By  Instagram Popularity Quotient  Fire Audit  \
0      Not Specific                           95.8           1   
1  Tier A Celebrity                           85.0           1   
2  Tier A Celebrity                           68.2           1   
3  Tier A Celebrity                           83.6           0   
4  Tier A Celebrity                           76.8           1   

   Liquor License Obtained  Situated in a Multi Complex  Dedicated Parking  \
0                        1                            1                  0   
1                        1                            1                  1   
2                        1                            1                  1   
3                        1                            1                  1   
4                        1                            1                  1   

   ...  Has_nigerian  Has_peruvian Has_polish Has_salvadorian  Has_sapnish  \
0  ...             0             0          0               0            0   
1  ...             0             0          0               0            0   
2  ...             0             0          0               0            0   
3  ...             1             0          0               0            0   
4  ...             0             0          0               0            0   

   Has_swedish  Has_thai  Has_tibetan  Has_turkish  Has_welsh  
0            0         0            0            0          0  
1            0         0            0            0          0  
2            0         0            1            0          0  
3            0         0            0            1          0  
4            0         0            0            0          0  

[5 rows x 63 columns]